### DISCUSSION TITLE
How to run TypeOrm migrations

### DISCUSSION THREAD

**Discussion by djschilling** [sidebase/sidebase#77] [Q&A] [answered]
Thanks for the project and the setup with TypeOrm.

How can i run TypeOrm migrations(https://orkhan.gitbook.io/typeorm/docs/migrations)?
The default project seems to not have them included.
How should i set them up?

**Comment by djschilling** [comment]
@BracketJohn Do you know this?

**Reply by BracketJohn** [reply]
Yes, checkout: https://typeorm.io/migrations#running-and-reverting-migrations

Does this resolve your question? (:

**Reply by djschilling** [reply]
No it does not. Where do i put the migrations? How do in handle environment variables so that the data source can be used in the migrations? Do you have a sidebase example with migrations and which command to use to actually run them?

**Comment by djschilling** [comment]
Here is what fails:

This is what my file in `server/db/connection.ts` looks like:

```
import { DataSource } from 'typeorm'
import { Association } from '~~/src/entity/association'



const AppDataSource = new DataSource({
  type: 'mysql',
  synchronize: false,
  logging: false,
  host: process.env.TYPEORM_HOST,
  port: Number(process.env.TYPEORM_PORT),
  username: process.env.TYPEORM_USERNAME,
  password: process.env.TYPEORM_PASSWORD,
  database: process.env.TYPEORM_DATABASE,
  entities: [Association],
})

const initialize = async () => {
  console.log('DB: Initializing DB connection')

  if (AppDataSource.isInitialized) {
    console.log('DB: Already initialized')
    return
  }

  try {
    await AppDataSource.initialize()
  } catch (error) {
    console.trace('DB: Failed to initialized database', error)
    throw error
  }

  console.log('DB: Successfully initialized database connection')
}

export { AppDataSource, initialize }

```

If i run `npx typeorm-ts-node-esm migration:run -d server/db/connection.ts` i get the following error:

```
Error during migration run:
Error: Unable to open file: "/Users/david/projects/churchtools-finder/server/db/connection.ts". Cannot use import statement outside a module
    at Function.loadDataSource (/Users/david/projects/churchtools-finder/node_modules/src/commands/CommandUtils.ts:22:19)
    at async Object.handler (/Users/david/projects/churchtools-finder/node_modules/src/commands/MigrationRunCommand.ts:42:26)
```

I also tried `node --require ts-node/register ./node_modules/typeorm/cli.js migration:run -d server/db/connection.ts` and `npx typeorm-ts-node-commonjs migration:run -d server/db/connection.ts` with the same result.

I also looked at this thread but it did not help me: https://github.com/typeorm/typeorm/issues/8810

**Reply by BracketJohn** [reply]
Hey! I think I know a fix for this: Try to run:
```sh
> npx vite-node ...
```

instead of `node ...` -> with this the error above should be resolved (:

**Reply by djschilling** [reply]
i tried it like this and still got the same error:

`npx vite-node ./node_modules/typeorm/cli.js -- migration:run -d server/db/connection.ts`

**Reply by BracketJohn** [reply]
Can you show the error?

**Reply by djschilling** [reply]
It is the same error as shown above:

```
npx vite-node ./node_modules/typeorm/cli.js -- migration:run -d server/db/connection.ts
Error during migration run:
Error: Unable to open file: "/Users/david/projects/churchtools-finder/server/db/connection.ts". Cannot use import statement outside a module
    at Function.loadDataSource (/Users/david/projects/churchtools-finder/src/commands/CommandUtils.ts:22:19)
    at Object.handler (/Users/david/projects/churchtools-finder/src/commands/MigrationRunCommand.ts:42:26)
```

**Reply by BracketJohn** [reply]
Can you provide the repo / full code so that I can check it out?

In general: We're moving this repo to prisma ORM soon (see #80), with that it should hopefully be much simpler (:

**Reply by djschilling** [reply]
Unfortunately i cannot provide the repo.
And i also cannot switch to prisma, it is kind of a big project and a switch would take a lot of time.

Did you try to run those commands in the current sidebase project yourself? It should behave the same there.

**Reply by BracketJohn** [reply]
Sorry @djschilling - we've since migrated TypeORM as the broader ecosystem is mainly using Prisma atm.

Although I personally am also a big fan of TypeORM, I think it's easier and more intuitive to work with Prisma for many. Consider asking your question in the TypeORM community, they should be more engaged  and interested inr esolving this problem ❤️ 

Sorry that I cannot really help you out here, nevertheless I hope that you have happy holidays / new years and to see you again soon 👋

**Reply by damgem** [reply]
To provide some context to that error:

- `Cannot use import statement outside a module`: the CLI uses `package.json` to figure out which packaging system to use (cjs / esm). Add `"type": "module"` to enable the usage of `import` (instead of `require`)
   - adding `"type": "module"` is not required when using `esbuild` to solve the next problem as this creates a single file that does not need to `import`

- The next problem that potentially arises is:
   ```
   Error: Unable to open file: "/path/to/file/specified/as/datasource.ts". Cannot find module '/some/other/file/that/gets/imported'
   ```
  - nuxt uses `esbuild` under the hood which enables you to import without specifiying the file extension (e.g. `import X from 'Y'` instead of `import X from 'Y.js')
  - the CLI uses node which requires a file extension for the import
  - using `esbuild` to create a single js file that does not `import` solves this

**Comment by damgem** [comment, marked as answer]
For documentation purposes I’ll link the answer I wrote to your issue on typeorm’s repo you created while I researched that for an intern project :slightly_smiling_face: https://github.com/typeorm/typeorm/issues/9645#issuecomment-1380508284

This is only a workaround but the gist is:
- use `esbuild` to compile the file that exports `DataSource`
- now specify the compiled file in the `-d` flag

You can prefix every cli command with that and then use the CLI as used.

Has this resolved your issue? if so: please mark this as the resolution :)
